### PR TITLE
Removed StandardEditor class

### LINKED ISSUE #113: What `StandardEditor` should be?
After https://github.com/ckeditor/ckeditor5-core/issues/110 the whole initialization of the editor will be moved to the `Editor` class. `StandardEditor` will be just a set of methods.
 
Because of it, we should consider changing it to a mixing or utils or... leave it as a subclass because what we have now might be fine.

### LINKED ISSUE #115: Get rid of StandardEditor, introduce EditorWithUI interface
We decided that we will replace `StandardEditor` with a set of utils.

 - `setData` and `getData` methods will become `DataInterface` mixin.
 - `updateEditorElement ` (renamed to `updateElement`), `loadDataFromEditorElement` (renamed to `loadDataFromElement`) and `element` property will become `ElementInterface` mixin.
 - `_attachToForm` will become utli.
 - `static create` can be removed.

Additionally `EditorWithUI` interface should be defined. It will  introduce only docs:
 - `editor.ui` property,
 - `editor#uiReady` event.

All concrete editors should implement this interface.

### LINKED ISSUE #303: How to document which editor interfaces are required by specific features?
This issue is based on https://github.com/ckeditor/ckeditor5/issues/302.

The [ckeditor5-core](https://github.com/ckeditor/ckeditor5-core) package is just a framework for creating RTEs. Final implementations, called also creators, are provided by other packages. Those creators extend base classes (atm it's `Editor` and `StandardEditor`) exposed by the core.

Now, on each level (`Editor` -> `StandardEditor` -> `ClassicEditor`) we've got a different set of properties available. E.g. `Editor` doesn't have `keystrokes` property because it's UI related. It appears in `StandardEditor` which is understood as "browser based editor with a single editable".

Most features are split into their "engine" and "full" parts. The [`BoldEngine`](https://github.com/ckeditor/ckeditor5-basic-styles/blob/master/src/boldengine.js) provides only the command, schema and converters. It's compatible with editors which have `document` and `commands` and `editing` properties, so by `Editor`'s children (there's a trick – `Editor`'s constructor doesn't set `editing` property but defines it and says that it must be set). However, [`Bold`](https://github.com/ckeditor/ckeditor5-basic-styles/blob/master/src/bold.js) requires much more – `keystrokes` and `ui`. It will work with `StandardEditor` and every other editor implementing the same interface.

What we're missing are definitions of these interfaces and features defining what kind of interface they need. Without them, end-developers won't know what kind of creator is needed for the set of features that he/she is using.

The easiest solution will be to create a set of standard interface definitions and in each feature's documentation define which interface must be provided.

We can also automate this by introducing new properties. E.g. `Editor` could have a `implements` property which is an array of strings (interface names) and feature `requiresEditor` (or something more intuitive) which defines which interfaces the editor must fulfill. If requirements are not met, we could throw an error.

### LINKED ISSUE #2930: Get rid of StandardEditor, introduce EditorWithUI interface
We decided that we will replace `StandardEditor` with a set of utils.

 - `setData` and `getData` methods will become `DataInterface` mixin.
 - `updateEditorElement ` (renamed to `updateElement`), `loadDataFromEditorElement` (renamed to `loadDataFromElement`) and `element` property will become `ElementInterface` mixin.
 - `_attachToForm` will become utli.
 - `static create` can be removed.

Additionally `EditorWithUI` interface should be defined. It will  introduce only docs:
 - `editor.ui` property,
 - `editor#uiReady` event.

All concrete editors should implement this interface.

### LINKED ISSUE #2931: What `StandardEditor` should be?
After https://github.com/ckeditor/ckeditor5-core/issues/110 the whole initialization of the editor will be moved to the `Editor` class. `StandardEditor` will be just a set of methods.
 
Because of it, we should consider changing it to a mixing or utils or... leave it as a subclass because what we have now might be fine.

### PR DESCRIPTION
**Pull request by oskarwrobel**
### Suggested merge commit message ([convention](https://github.com/ckeditor/ckeditor5-design/wiki/Git-commit-message-convention))

Fix: Removed `StandardEditor` class in favor of `DataInterface` and `ElementInterface` mixins. Added `EditorWithUI` interface. Closes ckeditor/ckeditor5#2930. Closes ckeditor/ckeditor5#2931. Closes https://github.com/ckeditor/ckeditor5/issues/303.

BREAKING CHANGE: `StandardEditor` class is removed. Use `Editor` class with `DataInterface` and `ElementInterface` mixins.

---

Constelation: https://github.com/ckeditor/ckeditor5/tree/t/ckeditor5-core/115

Requires:
- https://github.com/ckeditor/ckeditor5-editor-classic/tree/t/ckeditor5-core/115 (major changes)
- https://github.com/ckeditor/ckeditor5-editor-balloon/tree/t/ckeditor5-core/115 (major changes)
- https://github.com/ckeditor/ckeditor5-editor-inline/tree/t/ckeditor5-core/115 (major changes)
- https://github.com/ckeditor/ckeditor5-ui/tree/t/ckeditor5-core/115 (changes in docs)
- https://github.com/ckeditor/ckeditor5-markdown-gfm/tree/t/ckeditor5-core/115 (changes in docs)
- https://github.com/ckeditor/ckeditor5-engine/tree/t/ckeditor5-core/115 (minor changes in tests)
- https://github.com/ckeditor/ckeditor5/tree/t/ckeditor5-core/115b (changes in intro docs)

### CODE REVIEW

File: src/editor/editor.js [hunk 1]
```diff
@@ -21,10 +21,9 @@ import mix from '@ckeditor/ckeditor5-utils/src/mix';
 
 /**
  * Class representing the base of the editor. It is the API all plugins can expect to get when using editor property.
- * Editors implementation (like Classic Editor or Inline Editor) should extend this class. They can add their own
- * methods and properties.
- *
- * See also {@link module:core/editor/standardeditor~StandardEditor}.
+ * Editors implementation (like {@link module:editor-classic/classiceditor~ClassicEditor Classic Editor} or
```
Comment (Reinmar) [inline src/editor/editor.js]: "Classic editor" or simply remove this (so it will say `ClassicEditor`).

File: src/editor/utils/elementinterface.js [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+ * @implements module:core/editor/utils/elementinterface~Element
+ */
+const ElementInterface = {
+	/**
+	 * The element on which the editor has been initialized.
+	 *
+	 * @readonly
+	 * @member {HTMLElement} #element
+	 */
+	element: null,
```
Comment (Reinmar) [inline src/editor/utils/elementinterface.js]: This should not be a part of the mixin. It will go into the prototype anyway. It's simply unnecessary here.

File: src/editor/utils/datainterface.js [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+ */
+
+/**
+ * Interface provides methods for setting and getting data to/from editor main root element of the model tree
+ * using {@link module:core/editor/editor~Editor#data data pipeline}.
+ *
+ * @mixin DataInterface
+ * @implements module:core/editor/utils/datainterface~Data
+ */
+const DataInterface = {
```
Comment (Reinmar) [inline src/editor/utils/datainterface.js]: So far we called all mixins "mixins". Like `ObservableMixin` which implements the `Observable` interface. So we didn't use the word `Interface`.

Also, a simple `Data` interface is kinda confusing. Editor, by implementing the `Data` interface becomes a data itself? The name needs to be clearer.

I'd propose something like:

* `EditorDataApi` interface
* `EditorDataApiMixin` mixin

File: src/editor/utils/datainterface.js [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+	 * Sets the data in the editor's main root.
+	 *
+	 * @method #setData
+	 * @param {*} data The data to load.
+	 */
+	setData( data ) {
+		this.data.set( data );
+	},
+
+	/**
```
Comment (Reinmar) [inline src/editor/utils/datainterface.js]: Definition of these should go into the interface, not to the mixin. The interface defines the API. The mixin implements it.

File: src/editor/utils/elementinterface.js [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+ * @module core/editor/utils/elementinterface
+ */
+
+/**
+ * Interface provides method for setting and getting data from/to element on which editor has been initialized.
+ *
+ * @mixin ElementInterface
+ * @implements module:core/editor/utils/elementinterface~Element
+ */
+const ElementInterface = {
```
Comment (Reinmar) [inline src/editor/utils/elementinterface.js]: All the same notes as in case of the `EditorDataApi`. Here we could have:

* `EditorElementMixin` mixin
* `EditorElement` interface

But, just like with the "data" stuff, I don't like the fact that editor doesn't become an editor element but it gets an API of such type, so (however I don't like the length of this) the following seems more correct:

* `EditorElementApiMixin`
* `EditorElementApi`

PS. API ~= Interface, but I use API because we'd end up with `EditorElementInterfaceMixin` and `EditorElementInterface`. Unless... we'd decide that we can omit `Interface` suffix, but then we end up with an also ugly `EditorElement` interface.

In other words, I chose "API" because it reads well: "My editor implements the editor element API". While this makes no sense at all: "My editor implements editor element".

File: src/editor/utils/datainterface.js (hunk 3 above)
Comment (pjasiun) [reply src/editor/utils/datainterface.js]: I'm fine with `EditorDataApiMixin`. However, note that these are in the `editor/utils` file and will be imported in the editor implementation class, so I believe it will be clear for everybody who imports it that, use it or see it there that it's about `DataApi` for the editor, not for something else. This is why `DataApiMixin` is fine for me as well.

File: src/editor/utils/elementinterface.js (hunk 2 above)
Comment (pjasiun) [reply src/editor/utils/elementinterface.js]: > into the prototype anyway.

You mean to the interface?

File: src/editor/utils/datainterface.js (hunk 3 above)
Comment (Reinmar) [reply src/editor/utils/datainterface.js]: 👍  for `DataApiMixin`. This will be consistent with our approach to not duplicate module's path in class names.

File: src/editor/utils/elementinterface.js (hunk 2 above)
Comment (Reinmar) [reply src/editor/utils/elementinterface.js]: By "will go" I meant that this property will be added to `SomeEditor.prototype`. It makes no sense there. Properties (physically) should only be added in instances. The property documentation, though, should go to the interface, together with other docs.

**Comment by oskarwrobel** [comment]
Ok guys, I think everything is fixed.
